enhancements/cloud-integration/azure/azure-workload-identity.md

```diff
@@ -0,0 +1,358 @@
+
+All operators would need code changes similar to the sample below, introducing `azidentity.NewWorkloadIdentityCredential()` for procuring a credential as an alternative to `azidentity.NewClientSecretCredential()` for the current config secret.
+
+```go
+func getAuthorizerForResource(config Config) (autorest.Authorizer, error) {
+    ...
+
+	// Fallback to using tenant ID from env variable if not set.
+	if strings.TrimSpace(config.TenantID) == "" {
+		config.TenantID = os.Getenv("AZURE_TENANT_ID")
+		if strings.TrimSpace(config.TenantID) == "" {
+			return nil, errors.New("empty tenant ID")
+		}
+	}
+
+	// Fallback to using client ID from env variable if not set.
+	if strings.TrimSpace(config.ClientID) == "" {
+		config.ClientID = os.Getenv("AZURE_CLIENT_ID")
+		if strings.TrimSpace(config.ClientID) == "" {
+			return nil, errors.New("empty client ID")
+		}
+	}
+
+	// Fallback to using client secret from env variable if not set.
+	if strings.TrimSpace(config.ClientSecret) == "" {
+		config.ClientID = os.Getenv("AZURE_CLIENT_SECRET")
+		// Skip validation; fallback to token (below) if env variable is also not set.
+	}
+
+	var (
+		cred azcore.TokenCredential
+		err  error
+	)
+	if strings.TrimSpace(config.ClientSecret) == "" {
+		options := azidentity.WorkloadIdentityCredentialOptions{
+			ClientOptions: azcore.ClientOptions{
+				Cloud: cloudConfig,
+			},
+		}
+		cred, err = azidentity.NewWorkloadIdentityCredential(config.TenantID, config.ClientID, "/var/run/secrets/openshift/serviceaccount/token", &options)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		options := azidentity.ClientSecretCredentialOptions{
+			ClientOptions: azcore.ClientOptions{
+				Cloud: cloudConfig,
+			},
+		}
+		cred, err = azidentity.NewClientSecretCredential(config.TenantID, config.ClientID, config.ClientSecret, &options)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	scope := endpointToScope(config.Environment.TokenAudience)
+
+	// Use an adapter so azidentity in the Azure SDK can be used as
+	// Authorizer when calling the Azure Management Packages, which we
+	// currently use. Once the Azure SDK clients (found in /sdk) move to
+	// stable, we can update our clients and they will be able to use the
+	// creds directly without the authorizer. The schedule is here:
+	// https://azure.github.io/azure-sdk/releases/latest/index.html#go
+	authorizer := azidext.NewTokenCredentialAdapter(cred, []string{scope})
+
+	return authorizer, nil
+}
+```
+
+#### Mutating admission webhook
+
+CCO will deploy and lifecycle the [Azure Workload Identity mutating admission webhook](https://azure.github.io/azure-workload-identity/docs/installation/mutating-admission-webhook.html) on Azure clusters such that users can annotate workload `ServiceAccounts` with Managed Identity details necessary for creating clients. When the mutating admission webhook finds these annotations on a
```

Comment:
is CCO guaranteed to be run ahead of all operators that would make use of it?

Reply:
As written no core operators will rely on the webhook but I don't think we can guarantee that CCO will be running ahead of them.